documentation/python.py

```diff
@@ -907,47 +907,59 @@ def enclosing_module_for(state: State, name: str) -> Optional[str]:
+    """
+
     assert state.current_module
 
     # If not a string and not a module, try looking if its name-mapped type is
     # in our name map. If it is, and its enclosing module is as well, use that
     # to have name mapping correctly applied for it.
-    name = None
+    module_name = None
     if not isinstance(object, str) and not inspect.ismodule(object):
         name_candidate = map_name_prefix(state, extract_type(object))
         if name_candidate in state.name_map:
-            name = enclosing_module_for(state, name_candidate)
+            module_name = enclosing_module_for(state, name_candidate)
 
     # If the above didn't succeed, try other options
-    if not name:
-        # If it's a string, assume it's a parsed name that's already in the
-        # name map, find the leaf module name and add it
-        if isinstance(object, str):
+    if not module_name:
+        if object == 'os.PathLike':
+            module_name = 'os'
+
+        # If it's a string, assume it's a parsed module_name that's already in the
+        # module_name map, find the leaf module module_name and add it
+        elif isinstance(object, str):
             # We should get string names only for pybind11 types, nothing else
             # TODO er wait, what about unknown annotations? those probably
             #   shouldn't get here at all?
             assert state.config['PYBIND11_COMPATIBILITY']
-            name = enclosing_module_for(state, object)
+            module_name = enclosing_module_for(state, object)
             # If there's no enclosing module, it's a builtin, for which we add
             # no dependency. Given that str is passed only from pybind, all
             # referenced names should be either builtin or known.
-            if not name:
-                assert '.' not in object
-                return
+            if not module_name:
+                assert '.' not in object, "Unknown type {} in module {}".format(object, state.current_module)
 
         # If it's directly a module (such as `typing` or `enum` passed from
-        # certain parts of the codebase), apply name mapping to it
+        # certain parts of the codebase), apply module_name mapping to it.
         elif inspect.ismodule(object):
-            name = map_name_prefix(state, object.__name__)
+            module_name = map_name_prefix(state, object.__name__)
 
         # Otherwise it's a class/function/enum/..., extract module name from
-        # it, apply name mapping
+        # it, apply name mapping.
         else:
-            name = map_name_prefix(state, object.__module__)
+            module_name = map_name_prefix(state, object.__module__)
 
     # Add it only if it's not a module self-reference and if it's not builtins
-    if name != 'builtins' and name != state.current_module:
-        state.module_dependencies[state.current_module].add(name)
+    if module_name != 'builtins' and module_name != state.current_module and (module_name is not None):
```

Comment:
The `module_name is not None` part feels strange, I can't see how it is related to the other changes here, in particular how the newly added treatment of `os.PathLike` could result in a `None`. Same for the filtering of `result.params` below.

If those are actually fixing some bug I was not aware of, it's worth putting into a separate commit and with tests that reproduce such behavior.

Reply:
The `module_name is not None` change is to satisfy the type checker. I could switch it to an assert if you'd prefer. 

I considered a refactor of this whole block to structure it with less nesting and mutation logic, but decided against it. In general, it's best to avoid this pattern of `name = None ; overwrite name conditionally`

Reply:
Turns out that this case is totally being hit frequently (e.g., when `object == 'int'` [cringe variable naming, being that `object` describes all classes, also]). The value `None` _is_ making its way into that set, and just gets ignored later on. 

I propose keeping it exactly as I have it currently (with the `module_name is not None` check).